pennylane_forest/converter.py

```diff
@@ -0,0 +1,626 @@
+    def defined_gates(self):
+        """The custom gates defined in the pyquil Program.
+        
+        Returns:
+            List[Union[pyquil.quil.DefGate, pyquil.quil.DefPermutationGate]]: The gates defined in the pyquil Program
+        """
+        return self.program.defined_gates
+
+    @property
+    def defined_gate_names(self):
+        """The names of the custom gates defined in the pyquil Program.
+
+        Returns:
+            List[str]: The names of the gates defined in the pyquil Program
+        """
+        return self._defined_gate_names
+
+    @property
+    def declarations(self):
+        """The declarations in the pyquil Program.
+
+        Returns:
+            List[pyquil.quil.Declaration]: The declarations in the pyquil Program
+        """
+        return self._declarations
+
+    @property
+    def defined_variable_names(self):
+        """The names of the variables defined in the pyquil Program
+        
+        Returns:
+            List[str]: The names of the variables defined in the pyquil Program
+        """
+        return [declaration.name for declaration in self._declarations]
+
+    def _load_template(self):
+        """Load the template corresponding to the pyquil Program.
+        
+        Raises:
+            qml.DeviceError: When the import of a forked gate is attempted
+        """
+        self._parametrized_gates = []
+
+        for i, instruction in enumerate(self.program.instructions):
+            # Skip all statements that are not gates (RESET, MEASURE, PRAGMA, ...)
+            if not _is_gate(instruction):
+                if not _is_declaration(instruction):
+                    warnings.warn(
+                        "Instruction Nr. {} is not supported by PennyLane and was ignored: {}".format(
+                            i + 1, instruction
+                        )
+                    )
+
+                continue
+
+            # Rename for better readability
+            gate = instruction
+
+            if _is_forked(gate):
+                raise qml.DeviceError(
+                    "Forked gates can not be imported into PennyLane, as this functionality is not supported. "
+                    + "Instruction Nr. {}, {} was a forked gate.".format(i + 1, gate)
+                )
+
+            resolved_gate = _resolve_gate(gate)
+
+            # If the gate is a DefGate or not all CONTROLLED statements can be resolved
+            # we resort to QubitUnitary
+            if _is_controlled(resolved_gate) or self._is_defined_gate(resolved_gate):
+                pl_gate = ParametrizedQubitUnitary(self._resolve_gate_matrix(resolved_gate))
+            else:
+                pl_gate = pyquil_inv_operation_map[resolved_gate.name]
+
+            parametrized_gate = ParametrizedGate(
+                pl_gate, gate.qubits, gate.params, _is_inverted(gate)
+            )
+
+            self._parametrized_gates.append(parametrized_gate)
+
+    def template(self, variable_map={}, wires=None):
```

Comment:
It would be nice to have `variable_map= None` and then 
```python
if variable_map is None:
    variable_map = {}
```
This way the default argument is not mutable, which could potentially lead to unwanted behaviour.

Reply:
(If you do agree to make the change, the corresponding docstrings will have to be changed e.g. on line 528.)

Reply:
In which case could that lead to unexpected behaviour?

Reply:
Honestly, as the code is right now, I feel that it should be fine. However, once the code is refactored and the logic changes, there might be a case when it does matter.

[Here is a description on using mutable arguments with an example](https://docs.python-guide.org/writing/gotchas/).

Most of the times, it's the linter, that complains about it (e.g. in PyCharm).

Reply:
Okay, you got me on that one, Antal! I didn't know python behaved that weirdly weith default arguments. I'll change it.